[SYCL][CUDA] Initial CUDA backend support

Alexander-Johnston opened the pull request:
Initial support of CUDA backend bringing roughly 40% of SYCL 1.2.1 CTS conformance on devices that support CUDA 10.2.

fwyzard commented on sycl/include/CL/sycl/backend/cuda.hpp:
```diff
@@ -0,0 +1,30 @@
+//
+//===----------------------------------------------------------------------===//
+
+namespace cl {
```
```suggestion
#include <CL/sycl/detail/defines.hpp>

__SYCL_INLINE namespace cl {
```

should fix the build error
```
[  2%] Building CXX object tools/sycl/plugins/cuda/CMakeFiles/pi_cuda.dir/pi_cuda.cpp.o
In file included from /data/user/fwyzard/sycl/llvm/sycl/include/CL/sycl/detail/pi.hpp:13,
                 from /data/user/fwyzard/sycl/llvm/sycl/plugins/cuda/pi_cuda.cpp:10:
/data/user/fwyzard/sycl/llvm/sycl/include/CL/sycl/detail/common.hpp:25:25: error: inline namespace must be specified at initial definition
 __SYCL_INLINE namespace cl {
                         ^~
In file included from /data/user/fwyzard/sycl/llvm/sycl/plugins/cuda/pi_cuda.cpp:9:
/data/user/fwyzard/sycl/llvm/sycl/include/CL/sycl/backend/cuda.hpp:9:11: note: 'cl' defined here
 namespace cl {
           ^~
```
```

garimagu commented on sycl/include/CL/sycl/detail/pi_cuda.hpp:
```diff
@@ -0,0 +1,479 @@
+//===-- pi_cuda.hpp - CUDA Plugin -----------------------------------------===//
```
Can you move this file to  plugins/cuda/ ? It is the better place for it.

alexbatashev commented on sycl/source/context.cpp:
```diff
@@ -23,25 +23,31 @@
 namespace sycl {
-context::context(const async_handler &AsyncHandler)
-    : context(default_selector().select_device(), AsyncHandler) {}
+context::context(const async_handler &AsyncHandler, bool usePrimaryContext)
```
Nit:
```suggestion
context::context(const async_handler &AsyncHandler, bool UsePrimaryContext)
```

Same applies to the rest of the file.

alexbatashev commented on sycl/CMakeLists.txt:
```diff
@@ -190,7 +193,6 @@ add_subdirectory( source )
 # SYCL toolchain builds all components: compiler, libraries, headers, etc.
 add_custom_target( sycl-toolchain
   DEPENDS ${SYCL_RT_LIBS}
-          pi_opencl
```
Why removing OpenCL support?

Alexander-Johnston replied:
The pi_opencl dependency is added to sycl-toolchain in the opencl plugin cmake `sycl/plugins/opencl/CMakeLists.txt` in https://github.com/intel/llvm/commit/884459b63f44239f8e6453dc3f9ed0f2502b8fc3#diff-edef34e019c51a57812a69f7a111afdcR22

alexbatashev commented on sycl/include/CL/sycl/detail/cg.hpp:
```diff
@@ -141,6 +172,15 @@ class HostKernelBase {
+  std::function<void(cl::sycl::interop_handler)> MFunc;
+
+public:
+  InteropTask(std::function<void(cl::sycl::interop_handler)> Func)
```
```suggestion
  InteropTask(function_class<void(cl::sycl::interop_handler)> Func)
```

alexbatashev commented on sycl/include/CL/sycl/detail/queue_impl.hpp:
```diff
@@ -48,7 +48,7 @@ class queue_impl {
       : queue_impl(Device,
                    detail::getSyclObjImpl(
-                       context(createSyclObjFromImpl<device>(Device))),
+                       context(createSyclObjFromImpl<device>(Device), {}, true)),
```
Nit: `{}` and `true` seem to me some "magic" values. Can you please add comments here?

alexbatashev commented on sycl/include/CL/sycl/detail/scheduler/scheduler.hpp:
```diff
@@ -77,7 +77,7 @@ class Scheduler {
   // Returns an event which indicates when these nodes are completed and host
   // accessor is ready for using.
-  EventImplPtr addHostAccessor(Requirement *Req);
+  EventImplPtr addHostAccessor(Requirement *Req, const bool destructor = false);
```
Nit:
```suggestion
  EventImplPtr addHostAccessor(Requirement *Req, const bool Destructor = false);
```

alexbatashev commented on sycl/test/basic_tests/access_to_subset.cpp:
```diff
@@ -1,8 +1,9 @@
-// RUN: %clangxx -fsycl %s -o %t.out
+// RUN: %clangxx -fsycl -fsycl-targets=%sycl_triple %s -o %t.out
```
Can you please clarify if `-fsycl-targets` is a mandatory argument for user to provide now? If yes, wouldn't it be better if it had a default value?

Ruyk replied:
Is not, it is only required on the test so we can compile correctly for CUDA or SPIRV targets.

alexbatashev commented on sycl/include/CL/sycl/handler.hpp:
```diff
@@ -771,6 +771,15 @@ class handler {
 #endif
   }
 
+  // Similar to single_task, but passed lambda will be executed on host
```
Nit: Can you please make doxygen comment out of this?

alexbatashev commented on sycl/source/CMakeLists.txt:
```diff
@@ -80,6 +81,7 @@ set(SYCL_SOURCES
     "detail/usm/usm_dispatch.cpp"
     "detail/usm/usm_impl.cpp"
     "detail/util.cpp"
+    "cg.cpp"
```
The corresponding header file resides in `CL/sycl/detail/`. Have you considered putting `cg.cpp` to `source/detail`?

alexbatashev commented on sycl/source/detail/accessor_impl.cpp:
```diff
@@ -14,8 +14,9 @@ namespace sycl {
 
 AccessorImplHost::~AccessorImplHost() {
-  if (MBlockedCmd)
+  if (MBlockedCmd) {
```
AFAIK, it's common practice in LLVM not to wrap single line statement in curly brackets.

alexbatashev commented on sycl/include/CL/sycl/context.hpp:
```diff
@@ -33,7 +33,10 @@ class context {
+  /// @param useCUDAPrimaryContext is a bool determining whether to use the
+  ///        primary context in the CUDA backend.
+  explicit context(const async_handler &AsyncHandler = {},
+                   bool useCUDAPrimaryContext = false);
```
Nit:
```suggestion
                   bool UseCUDAPrimaryContext = false);
```

jbrodman commented on clang/lib/Driver/ToolChains/Cuda.cpp:
```diff
@@ -61,7 +61,7 @@ static CudaVersion ParseCudaVersionFile(llvm::StringRef V) {
   if (Major == 10 && Minor == 0)
     return CudaVersion::CUDA_100;
-  if (Major == 10 && Minor == 1)
+  if (Major == 10 && (Minor == 1 || Minor == 2))
```
ToT Clang lags behind on released CUDA versions.

Ruyk replied:
We are hoping upstream clang catches up with 10.2 support, this change allows to use it on the meantime. Their release notes and our testing shows 10.2 can work as if 10.1 and there are no regressions. Otherwise, the default installation of many people simply fails to compile any SYCL for CUDA application.

```suggestion
// Enable CUDA 10.2 toolkit acting as 10.1 until tip clang catches up
  if (Major == 10 && (Minor == 1 || Minor == 2))
```

Ruyk replied:
Actually, upstream llvm patch https://github.com/llvm/llvm-project/commit/12fefeef203ab4ef52d19bcdbd4180608a4deae1  (https://reviews.llvm.org/D73231)   is a better fix, so we won't need once this lands into the intel/llvm fork

Ruyk commented:
Thanks for the feedback, we are working in fixing some lit-testing regressions and rebasing on top of the latest changes.

AlexeySachkov commented on clang/tools/clc-spirv-mapping-generator/ClcSpirvMappingGenerator.cpp:
```diff
@@ -0,0 +1,682 @@
+    // function.
+    "read_imagei",  // conflicts with `read_imagef`
+    "read_imageui", // conflicts with `read_imagef`
+    "read_imageh",  // conflicts with `read_imagef`
```
BTW, this should be fixed by KhronosGroup/SPIRV-LLVM-Translator#408

AlexeySachkov commented on llvm-spirv/lib/SPIRV/OCL20ToSPIRV.cpp:
```diff
@@ -288,11 +288,31 @@ class OCL20ToSPIRV : public ModulePass, public InstVisitor<OCL20ToSPIRV> {
   Module *M;
   LLVMContext *Ctx;
   unsigned CLVer; /// OpenCL version as major*10+minor
+  unsigned CLLang; /// OpenCL language, see `spv::SourceLanguage`.
```
Please note that this repo contains a copy of KrhonosGroup/SPIRV-LLVM-Translator which is updated from time to time and usually, by simple copy-pasting the whole directory. So, there is a huge chance that these changes will be overwritten occasionally

Please at least extract these changes to a separate commit, so it can be re-applied at each sync, or (which is much better) commit them directly to KhronosGroup/SPIRV-LLVM-Translator

Alexander-Johnston replied:
We will try to commit these directly to the Khronos repository, but will need the changes here until it is merged on the Khronos repository and pulled back down for the CUDA backend to work.

For now the changes to llvm-spirv are separated into https://github.com/intel/llvm/pull/1091/commits/09b3b2e401f03c08f3422f9700d8814da08fc43f

fwyzard commented:
[codeplaysoftware:cuda-dev](https://github.com/codeplaysoftware/sycl-for-cuda/tree/cuda-dev) fails to build if configured with `-DBUILD_SHARED_LIBS=ON`:
```
Scanning dependencies of target prepare_builtins
[ 64%] Building CXX object tools/libclc/utils/CMakeFiles/prepare_builtins.dir/prepare-builtins.cpp.o
[ 64%] Linking CXX executable ../../../bin/prepare_builtins
CMakeFiles/prepare_builtins.dir/prepare-builtins.cpp.o: In function `main':
prepare-builtins.cpp:(.text.startup.main+0x100): undefined reference to `llvm::parseBitcodeFile(llvm::MemoryBufferRef, llvm::LLVMContext&)'
prepare-builtins.cpp:(.text.startup.main+0x46c): undefined reference to `llvm::errorToErrorCodeAndEmitErrors(llvm::LLVMContext&, llvm::Error)'
collect2: error: ld returned 1 exit status
make[3]: *** [tools/libclc/utils/CMakeFiles/prepare_builtins.dir/build.make:90: bin/prepare_builtins] Error 1
make[2]: *** [CMakeFiles/Makefile2:98983: tools/libclc/utils/CMakeFiles/prepare_builtins.dir/all] Error 2
make[1]: *** [CMakeFiles/Makefile2:99597: tools/sycl/CMakeFiles/sycl-toolchain.dir/rule] Error 2
make: *** [Makefile:25159: sycl-toolchain] Error 2
```

The [intel:sycl](https://github.com/intel/llvm/tree/sycl) branch does build fine.

vladimirlaz commented on buildbot/configure.py:
```diff
@@ -11,30 +11,49 @@ def do_configure(args):
+        "-DCMAKE_BUILD_TYPE={}".format(args.build_type),
+        "-DLLVM_ENABLE_ASSERTIONS={}".format(llvm_enable_assertions),
+        "-DLLVM_TARGETS_TO_BUILD={}".format(llvm_targets_to_build),
+        "-DLLVM_EXTERNAL_PROJECTS=sycl;llvm-spirv",
```
opencl-aot project is missed

Ruyk replied:
Is this needed for the overall project as well or just for the testing? Its not mentioned on the Getting Started Guide

alexbatashev replied:
@Ruyk opencl-aot is an optional tool to enable AOT compilation for SYCL. It is not a requirement for end users, but it is being tested in CI.

vladimirlaz commented on buildbot/configure.py:
```diff
@@ -11,30 +11,49 @@ def do_configure(args):
-    icd_loader_lib = ''
+    icd_loader_lib = os.path.join(args.obj_dir, "OpenCL-ICD-Loader", "build")
+    llvm_targets_to_build = 'X86'
+    llvm_enable_projects = 'clang;llvm-spirv;sycl'
```
opencl-aot project is missed

bader commented on libclc/generic/include/spirv/math/fmod.h:
```diff
@@ -0,0 +1,12 @@
+//
+//===----------------------------------------------------------------------===//
+
+#define __SPIRV_FUNCTION __spirv_ocl_fmod
```
I don't see definition of `__spirv_ocl_fmod` function. Is it a bug?

Alexander-Johnston replied:
There may not be a definition for `__spirv_ocl_fmod` yet. We haven't finished implementing all of the builtins yet, but have some internal patches with additional builtins in progress to raise later.

bader commented on clang/include/clang/Config/config.h.cmake:
```diff
@@ -80,6 +80,9 @@
 #cmakedefine01 CLANG_ENABLE_STATIC_ANALYZER
 
+/* Define if we have SYCL PI CUDA support */
+#cmakedefine SYCL_HAVE_PI_CUDA ${SYCL_HAVE_PI_CUDA}
```
Do we really need this define? Can we have "SYCL PI CUDA support" unconditionally?

Alexander-Johnston replied:
If we were to have PI CUDA support unconditionally the cuda toolchain will always be required for compilation. We decided to make it optional to allow people who only use the OpenCL plugin to compile the project without a cuda toolchain on their system.

Alexander-Johnston replied:
If we were to have PI CUDA support unconditionally the cuda toolchain will always be required for compilation. We decided to make it optional to allow people who only use the OpenCL plugin to compile the project without a cuda toolchain on their system.

bader replied:
According to my understanding we need CUDA toolchain to build CUDA plugin only.
Could you clarify why we should require CUDA toolchain to build the driver?
https://llvm.org/docs/CompileCudaWithLLVM.html - doesn't seem to require some custom driver.

Alexander-Johnston replied:
You are correct, we only need CUDA toolchain for building of the plugin, but we limit the valid SYCL triples in the clang driver based if PI CUDA support is available or not.

https://github.com/intel/llvm/pull/1091/files#diff-beaf25b0cdf8830dd4ea165404b00671R618
```cpp
static bool isValidSYCLTriple(llvm::Triple T) {
#ifdef SYCL_HAVE_PI_CUDA
  // NVPTX is valid for SYCL.
  if (T.isNVPTX())
    return true;
#endif
  // Check for invalid SYCL device triple values.
  // Non-SPIR arch.
  if (!T.isSPIR())
    return false;
  // SPIR arch, but has invalid SubArch for AOT.
  StringRef A(T.getArchName());
  if (T.getSubArch() == llvm::Triple::NoSubArch &&
      ((T.getArch() == llvm::Triple::spir && !A.equals("spir")) ||
       (T.getArch() == llvm::Triple::spir64 && !A.equals("spir64"))))
    return false;
  return true;
}
```

We can remove this limitation though and always allow nvptx triples for compilation, regardless of it the CUDA plugin is available.

bader replied:
+1 for removing.

fwyzard commented on sycl/include/CL/sycl/backend/cuda.hpp:
```diff
@@ -0,0 +1,32 @@
+
+#include <CL/sycl/detail/defines.hpp>
+
+__SYCL_INLINE namespace cl {
```
Following #1121, this needs to be changed to 
```suggestion
__SYCL_INLINE_NAMESPACE(cl) {
```

fwyzard replied:
see https://github.com/codeplaysoftware/sycl-for-cuda/pull/4

Ruyk commented:
We are currently working on fixing the default device selection. The current implementation does not take into account the multiple backends, and causes several lit tests to fail. We should have a fix soon.

Alexander-Johnston commented on llvm/CMakeLists.txt:
```diff
@@ -462,6 +462,10 @@ if( LLVM_USE_PERF )
   endif( NOT CMAKE_SYSTEM_NAME MATCHES "Linux" )
 endif( LLVM_USE_PERF )
 
+option(SYCL_BUILD_PI_CUDA
```
We ifdef in/out SYCL NVPTX support in the clang Driver based on this variable, which is why we put it in `llvm/CMakeLists.txt` rather than in `llvm/sycl/CMakeLists.txt`

bader replied:
The name of the option is confusing in this case. PI has nothing to do with the driver.
Why do we need ifdef out SYCL NVPTS support in the clang driver?

Alexander-Johnston replied:
We don't need to, but it helps prevent users accidentally compiling kernels to the wrong target for the plugins they have available.

bader replied:
Are you able to compile kernels for PTX w/o CUDA toolchain?

bader replied:
I so, we could use this for testing SYCL for CUDA compiler using cross-compilation on Intel HW.

Alexander-Johnston replied:
I believe clang will fail to produce PTX kernels without a CUDA toolchain installed.

Ruyk replied:
We could enable support for the SYCL cuda toolchain if the NVPTX target is added, so it has the same requirements as the CUDA support in llvm? I suggest we do that on a separate patch

bader replied:
Note, the code added under ifdef is not covered by CI system, so it might be broken.
It seems to me that in order to enable compilation for NVPTX target we need just a few changes. Am I wrong?

Alexander-Johnston commented:
@bader @smaslov-intel 

We've pushed some more fixes for the CUDA backend but are still finding issues relating to the new plugin interface introduced in https://github.com/intel/llvm/pull/1030.

We've found that the new plugin interface is unstable when multiple plugins are available. Though we've fixed some issues related to mismatching devices/platforms/plugins when doing info queries and device selection, we are still finding device selection and some of the plugin internals unstable. As a result we are still seeing some test failures. The most notable is basic_tests/queue.cpp, which visibly fails for us when trying to use the OpenCL or CUDA plugins.

We also found a number of tests failing silently by catching exceptions from the sycl runtime, but not erroring after they occur.

We've added a new target, check-sycl-cuda, for specifically attempting to test with CUDA devices by passing SYCL_BE=PI_CUDA. We've also added a matching flag to check-sycl to ensure a path to test OpenCL devices. Without SYCL_BE it is nondeterministic which device is chosen by the default_selector used by many tests when multiple plugins are available.

We have also been testing without SYCL_BUILD_PI_CUDA enabled or CUDA available on the system with success.

For now we can either
* merge and not enable testing for the CUDA backend while we work together on the new plugin interfaces stability with multiple plugins
* merge and enable testing for the CUDA backend, but mark the failures as warnings instead of failures until the new plugin interface is stable with multiple plugins

To enable building/testing of the CUDA backend you will need to add SYCL_BUILD_PI_CUDA=ON to the buildbot CMake invocation and check-sycl-cuda as a test target to the buildbots.

smaslov-intel commented:
> We've also added a matching flag to check-sycl to ensure a path to test OpenCL devices. Without SYCL_BE it is nondeterministic which device is chosen by the default_selector...

That sounds weird. Isn't the `useBackend` deterministic in preferring OpenCL when SYCL_BE is unset?
 
> merge and enable testing for the CUDA backend, but mark the failures as warnings...

I'd prefer it this way, and work further to stabilize the plugins mechanism (tagging @garimagu)

bader commented on buildbot/configure.py:
```diff
@@ -11,30 +11,49 @@ def do_configure(args):
+    llvm_enable_projects = 'clang;llvm-spirv;sycl;opencl-aot'
+    libclc_targets_to_build = ''
+    sycl_build_pi_cuda = 'OFF'
+    llvm_enable_assertions = 'OFF'
```
FYI -This is build breaking change.
```
D:\buildbot\Product_worker_intel\Build_PR_With_Lit_Win\llvm.src\sycl\source\detail\os_util.cpp(202): error C2220: warning treated as error - no 'object' file generated
D:\buildbot\Product_worker_intel\Build_PR_With_Lit_Win\llvm.src\sycl\source\detail\os_util.cpp(202): warning C4189: 'Ret': local variable is initialized but not referenced
```

bader commented:
FYI: https://github.com/intel/llvm/pull/1159 - might be useful for passing LIT tests.
@Alexander-Johnston, it looks like @againull merged PR with conflicting changes. Could you merge resolve merge conflicts, please?

bader commented on sycl/test/basic_tests/image.cpp:
```diff
@@ -1,8 +1,12 @@
 
+// TODO: No CUDA image support
+// TODO: ptxas fatal   : Unresolved extern function '_Z17__spirv_ImageReadIDv4_f14ocl_image2d_roDv2_iET_T0_T1_'
+// XFAIL: windows || cuda
```
```suggestion
// XFAIL: cuda
```

bader commented on sycl/test/basic_tests/image.cpp:
```diff
@@ -1,8 +1,12 @@
-// RUN: %CPU_RUN_PLACEHOLDER %t.out
+// RUN: %clangxx -fsycl -fsycl-targets=%sycl_triple %s -o %t.out
+// RUNx: env SYCL_DEVICE_TYPE=HOST %t.out
+// RUNx: %CPU_RUN_PLACEHOLDER %t.out
```
Why these are disabled?

Alexander-Johnston replied:
Good catch. I suspect this was done during development and we missed it when cleaning up.

bader commented on sycl/test/basic_tests/stream/stream.cpp:
```diff
@@ -1,8 +1,12 @@
 // RUN: %ACC_RUN_PLACEHOLDER %t.out %ACC_CHECK_PLACEHOLDER
+
+// TODO: ptxas fatal   : Unresolved extern function '_Z18__spirv_SignBitSetf'
+// XFAIL: windows || cuda
```
```suggestion
// XFAIL: cuda
```

bader commented on sycl/test/hier_par/hier_par_basic.cpp:
```diff
@@ -6,12 +6,15 @@
 // RUN: %ACC_RUN_PLACEHOLDER %t.out
 
+// TODO: ptxas fatal   : Unresolved extern function '__spirv_ControlBarrier'
+// XFAIL: windows || cuda
```
```suggestion
// XFAIL: cuda
```

bader commented on clang/test/Misc/nvptx.languageOptsOpenCL.cl:
```diff
@@ -28,17 +28,21 @@
-#ifdef cl_khr_int64_base_atomics
-#error "Incorrect cl_khr_int64_base_atomics define"
-#endif
+// Temporarily disabling the following two tests as a work around for the SYCL
```
Could you add TODO or FIXME here, please?
It will easier to find these later.

bader commented on sycl/CMakeLists.txt:
```diff
@@ -139,6 +139,9 @@ install(DIRECTORY ${OPENCL_INCLUDE}/CL
 # Configure SYCL version macro
 set(sycl_inc_dir ${CMAKE_CURRENT_SOURCE_DIR}/include)
+set(sycl_src_dir ${CMAKE_CURRENT_SOURCE_DIR}/source)
+set(sycl_detail_inc_dir ${CMAKE_CURRENT_SOURCE_DIR}/include/CL/sycl/detail)
```
Please, use paths relative to `sycl_inc_dir` to include headers from `include/CL/sycl/detail`.
```suggestion

```

bader replied:
Sorry, I was thinking about changing C++ code from
```c++
#include <header.hpp>
```
to
```c++
#include <CL/sycl/detail/header.hpp>
```

Let's make this refactoring later, otherwise we will never commit this PR.

Alexander-Johnston replied:
Oh I see what you mean. That would be better. I'll do this in a followup PR.

bader commented on sycl/CMakeLists.txt:
```diff
@@ -139,6 +139,9 @@ install(DIRECTORY ${OPENCL_INCLUDE}/CL
 set(sycl_inc_dir ${CMAKE_CURRENT_SOURCE_DIR}/include)
+set(sycl_src_dir ${CMAKE_CURRENT_SOURCE_DIR}/source)
+set(sycl_detail_inc_dir ${CMAKE_CURRENT_SOURCE_DIR}/include/CL/sycl/detail)
+set(sycl_detail_src_dir ${CMAKE_CURRENT_SOURCE_DIR}/source/detail)
```
```suggestion

```
Defined, but not used.

bader commented:
@Alexander-Johnston, I'm going to do more thorough code review post-commit. This PR is too big to review in one shot.

garimagu commented on sycl/source/detail/scheduler/commands.cpp:
```diff
@@ -930,9 +930,16 @@ cl_int ExecCGCommand::enqueueImp() {
         cl_mem MemArg = (cl_mem)AllocaCmd->getMemAllocation();
         Plugin.call<PiApiKind::piKernelSetArg>(Kernel, Arg.MIndex,
                                                sizeof(cl_mem), &MemArg);
+        Plugin.call<PiApiKind::piKernelSetArg>(Kernel, Arg.MIndex,
```
@bader @Alexander-Johnston Was this extra API added by mistake?
Is there a reason for adding it twice?
FYI @rbegam .

jbrodman commented on sycl/source/detail/context_impl.cpp:
```diff
@@ -21,24 +23,40 @@ __SYCL_INLINE_NAMESPACE(cl) {
 
 context_impl::context_impl(const vector_class<cl::sycl::device> Devices,
-                           async_handler AsyncHandler)
+                           async_handler AsyncHandler, bool UseCUDAPrimaryContext)
```
What is the purpose of this variable?  It is unused.
